### PR TITLE
Implement streaming for GoogleVertexAI text and chat models

### LINKED ISSUE #2594: Add support for Vertex AI Generative AI streaming response
REST responses can now provide a stream tokens as they are generated. (See https://cloud.google.com/vertex-ai/docs/generative-ai/learn/streaming#text)

This involves:
- A JSON reply with the contents of the response being in an array, and the objects in this array streamed out
- We will need something that can take the streamed JSON and process the relevant objects as they come in
  - Perhaps something like https://www.npmjs.com/package/stream-json
  - However... this means another (optionally) dependent package

### PR DESCRIPTION
**Pull request by afirstenberg**
Fixes #2594 
Obsoletes #2625 (which can be abandoned and deleted).

The parser here is... probably not the most efficient, but should be efficient enough for the purposes. And it doesn't require any new outside libraries.

### CODE REVIEW

File: langchain/src/chat_models/googlevertexai/common.ts [hunk 1]
```diff
@@ -235,7 +276,9 @@ export class BaseChatGoogleVertexAI<AuthOptions>
     );
 
     const generations =
-      result?.data?.predictions?.map((prediction) =>
+      (
+        result?.data as GoogleVertexAILLMPredictions<GoogleVertexAIChatPrediction>
```
Comment (jacoblee93) [inline langchain/src/chat_models/googlevertexai/common.ts:280]: Is the cast necessary? Should be `any` anyway and if we wanted to be sure we should use a type guard?

File: langchain/src/embeddings/googlevertexai.ts [hunk 2]
```diff
@@ -114,9 +115,9 @@ export class GoogleVertexAIEmbeddings
       responses
         ?.map(
           (response) =>
-            response.data?.predictions?.map(
-              (result) => result.embeddings.values
-            ) ?? []
+            (
+              response?.data as GoogleVertexAILLMPredictions<GoogleVertexEmbeddingsResults>
```
Comment (jacoblee93) [inline langchain/src/embeddings/googlevertexai.ts:119]: See above

File: langchain/src/experimental/multimodal_embeddings/googlevertexai.ts [hunk 3]
```diff
@@ -127,7 +128,9 @@ export class GoogleVertexAIMultimodalEmbeddings
   responseToEmbeddings(
     response: GoogleVertexAILLMResponse<GoogleVertexAIMultimodalEmbeddingsResults>
   ): MediaEmbeddings[] {
-    return response.data.predictions.map((r) => ({
+    return (
+      response?.data as GoogleVertexAILLMPredictions<GoogleVertexAIMultimodalEmbeddingsResults>
```
Comment (jacoblee93) [inline langchain/src/experimental/multimodal_embeddings/googlevertexai.ts:132]: See above

File: langchain/src/chat_models/googlevertexai/common.ts [hunk 4]
```diff
@@ -214,19 +228,46 @@ export class BaseChatGoogleVertexAI<AuthOptions>
...
+    // that is either available or added to the queue
+    while (!stream.streamDone) {
+      const output = await stream.nextChunk();
+      const chunk =
+        output !== null
+          ? BaseChatGoogleVertexAI.convertPredictionChunk(output)
+          : new ChatGenerationChunk({
+              text: "",
+              message: new AIMessageChunk(""),
+              generationInfo: { finishReason: "stop" },
```
Comment (jacoblee93) [inline langchain/src/chat_models/googlevertexai/common.ts:259]: We should break in this case?

File: langchain/src/chat_models/googlevertexai/common.ts [hunk 5]
```diff
@@ -214,19 +228,46 @@ export class BaseChatGoogleVertexAI<AuthOptions>
...
+      _options
+    );
+
+    // Get the streaming parser of the response
+    const stream = result.data as GoogleVertexAIStream;
+
+    // Loop until the end of the stream
+    // During the loop, yield each time we get a chunk from the streaming parser
+    // that is either available or added to the queue
+    while (!stream.streamDone) {
```
Comment (jacoblee93) [inline langchain/src/chat_models/googlevertexai/common.ts:251]: Feels like this should be an iterator?

File: langchain/src/util/googlevertexai-connection.ts [hunk 6]
```diff
@@ -106,29 +191,229 @@ export class GoogleVertexAILLMConnection<
...
+  }
+
+  /**
+   * Is the stream done?
+   * A stream is only done if all of the following are true:
+   * - There is no more data to be added to the text buffer
+   * - There is no more data in the text buffer
+   * - There are no chunks that are waiting to be consumed
+   */
+  get streamDone(): boolean {
```
Comment (jacoblee93) [inline langchain/src/util/googlevertexai-connection.ts:411]: Would be awesome to implement as an iterator - I can take a crack at it later this week

File: langchain/src/util/googlevertexai-webauth.ts [hunk 7]
```diff
@@ -63,6 +96,7 @@ export class WebGoogleAuth implements GoogleAbstractedClient {
     const res = await fetch(opts.url, fetchOptions);
 
     if (!res.ok) {
+      console.error(res);
```
Comment (jacoblee93) [inline langchain/src/util/googlevertexai-webauth.ts:99]: Remove log?

File: langchain/src/util/googlevertexai-connection.ts [hunk 8]
```diff
@@ -106,29 +191,229 @@ export class GoogleVertexAILLMConnection<
...
+  async nextChunk(): Promise<any> {
+    if (this._chunkQueue.length > 0) {
+      // If there is data in the queue, return the next queue chunk
+      return this._chunkQueue.shift() as GenerationChunk;
+    } else {
+      // Otherwise, set up a promise that handleChunk will cause to be resolved
+      this._chunkPending = new Promise((resolve) => {
+        this._chunkResolution = resolve;
+      });
+      return await this._chunkPending;
```
Comment (jacoblee93) [inline langchain/src/util/googlevertexai-connection.ts:400]: Redundant `await`

**Review by jacoblee93 — COMMENTED** [review]
Looks awesome overall! See comments

File: langchain/src/chat_models/googlevertexai/common.ts (hunk 4 above)
Comment (afirstenberg) [reply langchain/src/chat_models/googlevertexai/common.ts:259]: I did it this way to make sure we have a clear final chunk, but am not opposed to just breaking in this case and not sending it.

File: langchain/src/chat_models/googlevertexai/common.ts (hunk 1 above)
Comment (afirstenberg) [reply langchain/src/chat_models/googlevertexai/common.ts:280]: This seemed cleaner (and more obvious) than using an any in the type for data and marking where to ignore it. (And ditto for all the other places where it happens.)

File: langchain/src/chat_models/googlevertexai/common.ts (hunk 5 above)
Comment (afirstenberg) [reply langchain/src/chat_models/googlevertexai/common.ts:251]: I had enough trouble trying to wrap my head around implementing the async iterator correctly. I don't think I wanted to make multiple, nested, async iterators.

File: langchain/src/util/googlevertexai-webauth.ts (hunk 7 above)
Comment (afirstenberg) [reply langchain/src/util/googlevertexai-webauth.ts:99]: Oops. Yeah. This was left in while I was debugging why my auth with web wasn't working.

File: langchain/src/util/googlevertexai-connection.ts (hunk 8 above)
Comment (afirstenberg) [reply langchain/src/util/googlevertexai-connection.ts:400]: Mmm! Good point and good catch.

**Comment by afirstenberg** [comment]
I shy away from making it an iterator for the reason I outlined above - this is more straightforward, and trying to wrap my head around async iterators that worked with async iterators is making my head spin.
Is there a reason in favor of them in this case?

File: langchain/src/chat_models/googlevertexai/common.ts (hunk 4 above)
Comment (jacoblee93) [reply langchain/src/chat_models/googlevertexai/common.ts:259]: Oh, no I mean break after getting `null`. It's fine with me to have the final chunk like this but we wouldn't want to receive any more.

**Comment by jacoblee93** [comment]
Yeah they are tricky to implement, benefits are the standardized `for await const` loop consumption flow and the methods we have to convert them to readable streams, but I think not necessary in this case since they're not user facing.

**Comment by jacoblee93** [comment]
Hey @afirstenberg actually this seems to have broken the web integration test for non-streaming:

```typescript
MacBook-Pro-2:langchainjs jacoblee$ yarn test:single langchain/src/chat_models/tests/chatgooglevertexai_web.int.test.ts
(node:20994) ExperimentalWarning: VM Modules is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
  console.log
    {
      "generations": [
        []
      ],
      "llmOutput": []
    }

      at Object.log (src/chat_models/tests/chatgooglevertexai_web.int.test.ts:39:13)

  console.log
    [ [] ]

      at Object.log (src/chat_models/tests/chatgooglevertexai_web.int.test.ts:68:13)

  console.log
    [ [] ]

      at Object.log (src/chat_models/tests/chatgooglevertexai_web.int.test.ts:86:13)

  console.log
    [
     []
    ]
```

**Comment by jacoblee93** [comment]
Fixed it, should be good to go

File: langchain/src/chat_models/googlevertexai/common.ts (hunk 4 above)
Comment (afirstenberg) [reply langchain/src/chat_models/googlevertexai/common.ts:259]: We need to yield with that chunk, and after that it should be the last thing in the queue or that gets resolved, so the while loop will terminate.

**Pull request merged**